File: src/checks/rules/comments/shell.ts
import type { CommentRange } from "./ranges.js";

type ShellHeredoc = {
  delimiter: string;
  stripTabs: boolean;
};

function isShellCommentBoundary(line: string, index: number): boolean {
  if (index === 0) return true;
  return /[\s;&|(){}]/u.test(line[index - 1] ?? "");
}

function shellCommentStart(line: string): number | null {
  let single = false;
  let double = false;
  let backtick = false;
  let escaped = false;
  for (let index = 0; index < line.length; index += 1) {
    const char = line[index] ?? "";
    if (escaped) {
      escaped = false;
      continue;
    }
    if (!single && char === "\\") {
      escaped = true;
      continue;
    }
    if (!double && !backtick && char === "'") single = !single;
    else if (!single && !backtick && char === "\"") double = !double;
    else if (!single && char === "`") backtick = !backtick;
    else if (!single && !double && !backtick && char === "#" && isShellCommentBoundary(line, index)) return index;
  }
  return null;
}

function parseShellHeredocWord(line: string, start: number, limit: number): string {
  const quote = line[start];
  if (quote === "\"" || quote === "'") {
    const end = line.indexOf(quote, start + 1);
    return end > start && end <= limit ? line.slice(start + 1, end) : "";
  }
  let end = start;
  while (end < limit && !/[\s;&|()<>]/u.test(line[end] ?? "")) end += 1;
  return line.slice(start, end);
}

function shellHeredocDelimiter(line: string, limit: number): ShellHeredoc | null {
  let single = false;
  let double = false;
  for (let index = 0; index < limit - 1; index += 1) {
    const char = line[index] ?? "";
    if (!double && char === "'") single = !single;
    else if (!single && char === "\"") double = !double;
    if (single || double || char !== "<" || line[index + 1] !== "<") continue;
    let cursor = index + 2;
    const stripTabs = line[cursor] === "-";
    if (stripTabs) cursor += 1;
    while (cursor < limit && /\s/u.test(line[cursor] ?? "")) cursor += 1;
    const delimiter = parseShellHeredocWord(line, cursor, limit);
    if (delimiter) return { delimiter, stripTabs };
  }
  return null;
}

function collectShellCommentRanges(text: string): CommentRange[] {
  const ranges: CommentRange[] = [];
  let lineStart = 0;
  let heredoc: ShellHeredoc | null = null;
  while (lineStart <= text.length) {
    const newline = text.indexOf("\n", lineStart);
    const contentEnd = newline > 0 && text[newline - 1] === "\r" ? newline - 1 : newline < 0 ? text.length : newline;
    const line = text.slice(lineStart, contentEnd);
    if (heredoc) {
      if ((heredoc.stripTabs ? line.replace(/^\t+/u, "") : line).trimEnd() === heredoc.delimiter) heredoc = null;
    } else {
      const commentStart = shellCommentStart(line);
      const limit = commentStart ?? line.length;
      if (!(lineStart === 0 && line.startsWith("#!")) && commentStart != null) {
        ranges.push({ start: lineStart + commentStart, end: contentEnd, kind: "line" });
      }
      heredoc = shellHeredocDelimiter(line, limit);
    }
    if (newline < 0) break;
    lineStart = newline + 1;
  }
  return ranges;
}

export { collectShellCommentRanges };

File: native/code-discipline-native/src/parts/comment/ranges/shell.rs
fn is_shell_comment_boundary(line: &str, index: usize) -> bool {
    if index == 0 {
        return true;
    }
    line.as_bytes()
    .get(index - 1)
    .is_some_and(|byte| byte.is_ascii_whitespace() || b";&|(){}".contains(byte))
}

fn shell_comment_start(line: &str) -> Option<usize> {
    let bytes = line.as_bytes();
    let mut single = false;
    let mut double = false;
    let mut backtick = false;
    let mut escaped = false;

    for index in 0..bytes.len() {
        let current = bytes[index];
        if escaped {
            escaped = false;
            continue;
        }
        if !single && current == b'\\' {
            escaped = true;
            continue;
        }
        if !double && !backtick && current == b'\'' {
            single = !single;
        } else if !single && !backtick && current == b'"' {
            double = !double;
        } else if !single && current == b'`' {
            backtick = !backtick;
        } else if !single
        && !double
        && !backtick
        && current == b'#'
        && is_shell_comment_boundary(line, index)
        {
            return Some(index);
        }
    }

    None
}

fn parse_shell_heredoc_word(line: &str, start: usize, limit: usize) -> String {
    let bytes = line.as_bytes();
    if matches!(bytes.get(start), Some(b'"') | Some(b'\'')) {
        let quote = bytes[start];
        let mut end = start + 1;
        while end < limit && bytes.get(end) != Some(&quote) {
            end += 1;
        }
        return if end < limit {
            line[start + 1..end].to_string()
        } else {
            String::new()
        };
    }
    let mut end = start;
    while end < limit && !bytes[end].is_ascii_whitespace() && !b";&|()<>".contains(&bytes[end])
    {
        end += 1;
    }
    line[start..end].to_string()
}

fn shell_heredoc_delimiter(line: &str, limit: usize) -> Option<(String, bool)> {
    let bytes = line.as_bytes();
    let mut single = false;
    let mut double = false;
    let mut index = 0_usize;
    while index + 1 < limit {
        let current = bytes[index];
        if !double && current == b'\'' {
            single = !single;
        } else if !single && current == b'"' {
            double = !double;
        }
        if single || double || current != b'<' || bytes.get(index + 1) != Some(&b'<') {
            index += 1;
            continue;
        }
        let mut cursor = index + 2;
        let strip_tabs = bytes.get(cursor) == Some(&b'-');
        if strip_tabs {
            cursor += 1;
        }
        while cursor < limit && bytes[cursor].is_ascii_whitespace() {
            cursor += 1;
        }
        let delimiter = parse_shell_heredoc_word(line, cursor, limit);
        return if delimiter.is_empty() {
            None
        } else {
            Some((delimiter, strip_tabs))
        };
    }
    None
}

fn collect_shell_comment_ranges(text: &str) -> Vec<CommentRange> {
    let mut ranges = Vec::new();
    let mut line_start = 0_usize;
    let mut heredoc: Option<(String, bool)> = None;

    while line_start <= text.len() {
        let newline = text[line_start..].find('\n').map(|index| line_start + index);
        let content_end = match newline {
            Some(index) if index > 0 && text.as_bytes()[index - 1] == b'\r' => index - 1,
            Some(index) => index,
            None => text.len(),
        };
        let line = &text[line_start..content_end];
        if let Some((delimiter, strip_tabs)) = heredoc.as_ref() {
            let candidate = if *strip_tabs {
                line.trim_start_matches('\t')
            } else {
                line
            };
            if candidate.trim_end() == delimiter {
                heredoc = None;
            }
        } else {
            let comment_start = shell_comment_start(line);
            let limit = comment_start.unwrap_or(line.len());
            if !(line_start == 0 && line.starts_with("#!")) {
                if let Some(start) = comment_start {
                    ranges.push(CommentRange {
                            start: line_start + start,
                            end: content_end,
                            kind: CommentKind::Line,
                    });
                }
            }
            heredoc = shell_heredoc_delimiter(line, limit);
        }
        let Some(newline_index) = newline else {
            break;
        };
        line_start = newline_index + 1;
    }

    ranges
}
